files/routes/chat.py

```diff
@@ -1,117 +1,271 @@
+import functools
 import time
-from files.helpers.config.environment import SITE, SITE_FULL
-from files.helpers.wrappers import auth_required
-from files.helpers.sanitize import sanitize
+import uuid
+from typing import Any, Final
+
+from flask_socketio import SocketIO, emit, disconnect
+
+from files.__main__ import app, cache, limiter
+from files.helpers.alerts import *
 from files.helpers.config.const import *
-from datetime import datetime
-from flask_socketio import SocketIO, emit
-from files.__main__ import app, limiter, cache
-from flask import render_template, make_response, send_from_directory
-import sys
-import atexit
-
-if SITE == 'localhost':
-	socketio = SocketIO(app, async_mode='gevent', cors_allowed_origins=[SITE_FULL], logger=True, engineio_logger=True, debug=True)
+from files.helpers.config.environment import *
+from files.helpers.config.regex import *
+from files.helpers.sanitize import sanitize
+from files.helpers.wrappers import get_logged_in_user, is_not_permabanned, admin_level_required
+
+def chat_is_allowed(perm_level: int=0):
+	def wrapper_maker(func):
+		@functools.wraps(func)
+		def wrapper(*args: Any, **kwargs: Any) -> bool | None:
+			v = get_logged_in_user()
+			if not v:
+				return abort(403)
+			if not v.can_access_chat:
+				return abort(403)
+			if v.admin_level < perm_level:
+				return abort(403)
+			kwargs['v'] = v
+			return func(*args, **kwargs)
+		return wrapper
+	return wrapper_maker
+
+commands = {}
+
+def register_command(cmd_name, permission_level = 0):
+	def decorator(func):
+		@functools.wraps(func)
+		def wrapper(*args: Any, **kwargs: Any) -> bool | None:
+			v = get_logged_in_user()
+			if v.admin_level < permission_level:
+				send_system_reply(f"Unknown command: {cmd_name}")
```

Comment:
I don't think being paranoid is justified here. Also, what's up with not checking that v is not None as above?

Reply:
Actually, this one *is* necessary; these commands are called from the SPEAK handler. The Speak handler does the standard verification that the user is logged in and has access to chat, so we don't need to verify that the user is logged in (they are!), but all the verification we have there is to ensure that the user has chat access.

If we want to check if they have admin access, we need to do it here, in what is basically a nested handler.

Reply:
I meant, why are you returning a wrong error message. Tell 'em they are not authorized.

Reply:
Eh, I'm just overly paranoid in general. Doesn't seem like it causes a problem here, given that the only commands that exist are admin-only.